clang/test/SemaSYCL/sycl-device.cpp

```diff
@@ -21,13 +21,21 @@ namespace {
 }
 
 class A {
-  __attribute__((sycl_device)) // expected-error {{'sycl_device' attribute cannot be applied to a class member function}}
+  __attribute__((sycl_device))
   A() {}
 
-  __attribute__((sycl_device)) // expected-error {{'sycl_device' attribute cannot be applied to a class member function}}
+  __attribute__((sycl_device))
   int func3() {}
 };
 
+class B {
+  __attribute__((sycl_device)) // expected-error {{'sycl_device' attribute cannot be applied to a virtual or pure virtual class member function}}
+  virtual int foo() {}
+
+  __attribute__((sycl_device)) // expected-error {{'sycl_device' attribute cannot be applied to a virtual or pure virtual class member function}}
+  virtual int bar() = 0;
```

Comment:
Add a test for override/final as well!  Additionally, cases where the class itself is 'final'.  

I also wonder if we should ALLOW final versions of the function (or when the class is marked final).

Reply:
I suppose, technically we could allow any stuff which do not bring virtual table usage into resulting LLVM IR module.
I had some interesting observation:
First, any call of virtual function doesn't bring vtable usage to IR if it was done without using of pointer to the base class, i.e.
```
struct Bar {            
  virtual void bar1() {}
};
__attribute__((sycl_device)) void FOO() {
Bar b;   
b.bar1();
}
```
But, such calls will be diagnosed by our diagnostic in SemaSYCL here https://github.com/intel/llvm/blob/092367be7785a7764c6dc41497f2c74c6e164824/clang/lib/Sema/SemaSYCL.cpp#L226 .

Second, the diagnostic in SemaSYCL always triggered by the functions marked as `override` and `final` if they were marked as virtual in base class.
BTW if you are using function marked as **final** but through pointer to the base class, vtable usage will appear in resulting device code module anyway.
Another funny thing that the diagnostic implemented in place where attribute handling happens i.e. inside `handleSYCLDeviceAttr` function triggers only if function is explicitly marked as virtual, so, if you, for example, just override some virtual function and do not leave `virtual` keyword near it's declaration/definition, it will be possible to add `sycl_kernel` attribute to it, but calling of this function will be prohibited by the SemaSYCL diagnostic. I guess this happens because SemaSYCL diagnostic happens much more later when more information is presented. 

The point to disallow `sycl_kernel` attribute for virtual function was to improve user experience, I guess, because virtual functions cannot be called on device, which makes marking them with `sycl_device` useless.
But I suppose if this diagnostic is not so efficient and it's behavior can't be aligned with the diagnostic which validates actual virtual function calls, we can do not implement it to do not confuse users. Because now it looks like this:
```
struct Bar {                                       
  virtual void bar1() {}                           
};                                                 
                                                   
struct Bar1 : Bar {                                
  __attribute__((sycl_device)) void bar1() final {} // no diagnostic
};                                                 
                                                   
__attribute__((sycl_device)) void FOO() {          
  Bar1 b;                                          
  b.bar1();  // sycl-device.cpp:65:5: error: SYCL kernel cannot call a virtual function
                 // b.bar1();                                                          
}                                        
```
As a conclusion, I'm asking. @AlexeySachkov , @erichkeane is it ok if we won't implement diagnostic which disallow usage of `sycl_kernel`/`indirectly_callable` attirubute for virtual functions.

Please NOTE: my observation about that diagnostic for virtual function calls triggers even for cases which do not bring vtable usage to device code is not in scope of this PR.

Reply:
Hmm... yeah, I had forgotten about the parts where we already know the type.

    Bar &b =...;
    b.bar1();

Should definitely diagnose.  However, any non-ref/pointer types already know their type.  Anything marked 'override' should still obviously diagnose, BUT, this should not:

    Bar1 &b=...;
    b.bar1();

Because bar1 is final in Bar1, I would expect us to KNOW the call and not need to use the vtable ( though I am not sure we CAN do this, I'm not sure if this sitll uses the vtable).

Another example:

    struct Bar2 : Bar {
        __attribute___((sycl_device)) void bar1() override {} /// should still diagnose.

Reply:
As I said, problems with diagnosing of virtual function calls is not in scope of this PR. I just captured my observations while adding new test cases and exploring behavior.
This PR is about `SYCL_EXTERNAL` and therefore about `__attribute___((sycl_device))`
Such case can't be diagnosed in a way I use now:
```
struct Bar2 : Bar {
    __attribute___((sycl_device)) void bar1() override {}
```
Because inside `handleSYCLDeviceAttr` function we don't have enough information yet, so here diagnostic happens only if function explicitly marked with `virtual` keyword. But once this function is called, the call will be diagnosed by our old diagnostic from SemaSYCL because it happens later with more information.
So, I propose do not diagnose `sycl_device` usage on virtual functions, because such diagnostic is not sufficient.

Reply:
Right, I'm exploring the problem space here... I don't think skipping this diagnostic is a bad idea if we can properly count on diagnosing calls.